crates/sui-types/src/base_types.rs

```diff
@@ -653,22 +654,29 @@ impl SequenceNumber {
         SequenceNumber(u)
     }
 
+    pub fn increment_to(&mut self, next: SequenceNumber) {
+        debug_assert!(*self < next, "Not an increment: {} to {}", self, next);
+        *self = next;
+    }
+
+    pub fn decrement_to(&mut self, prev: SequenceNumber) {
+        debug_assert!(prev < *self, "Not a decrement: {} to {}", self, prev);
+        *self = prev;
+    }
+
+    /// Returns a new sequence number that is greater than all `SequenceNumber`s in `inputs`,
+    /// assuming this operation will not overflow.
     #[must_use]
-    pub fn increment(self) -> SequenceNumber {
-        // TODO: Ensure this never overflow.
+    pub fn lamport_increment(inputs: impl IntoIterator<Item = SequenceNumber>) -> SequenceNumber {
+        let max_input = inputs.into_iter().fold(SequenceNumber::new(), max);
+
+        // TODO: Ensure this never overflows.
         // Option 1: Freeze the object when sequence number reaches MAX.
         // Option 2: Reject tx with MAX sequence number.
         // Issue #182.
-        debug_assert_ne!(self.0, u64::MAX);
-        Self(self.0 + 1)
-    }
+        debug_assert_ne!(max_input.0, u64::MAX);
```

Comment:
this should be a regular assert - we can't reasonably continue if we know the following addition will overflow.